func-complete-check/check-marker/check-marker.v1.ts

```diff
@@ -0,0 +1,166 @@
+import * as RA from 'ramda-adjunct'
+import * as R from 'ramda'
+import { IAsyncTableService, AsyncTableService } from '../lib/storage-helper'
+import { ValidatedCheck } from '../typings/message-schemas'
+import moment from 'moment'
+import { ICheckFormService, CheckFormService } from '../lib/check-form.service'
+
+export interface ICheckMarkerFunctionBindings {
+  receivedCheckTable: Array<any>
+  checkNotificationQueue: Array<any>
+}
+
+declare interface Answer {
+  factor1: number
+  factor2: number
+  answer: string
+  sequenceNumber: number
+  question: string
+  clientTimestamp: string
+}
+
+declare interface CheckFormQuestion {
+  f1: number
+  f2: number
+}
+
+declare interface MarkingData {
+  answers: Array<Answer>
+  formQuestions: Array<CheckFormQuestion>
+  results: any
+}
+
+declare interface Mark {
+  mark: number
+  maxMarks: number
+  processedAt: Date
+}
+
+export class CheckMarkerV1 {
+
+  private _tableService: IAsyncTableService
+  private _sqlService: ICheckFormService
+
+  constructor (tableService?: IAsyncTableService, sqlService?: ICheckFormService) {
+    if (tableService === undefined) {
+      this._tableService = new AsyncTableService()
+    } else {
+      this._tableService = tableService
+    }
+
+    if (sqlService === undefined) {
+      this._sqlService = new CheckFormService()
+    } else {
+      this._sqlService = sqlService
+    }
+  }
+
+  async mark (functionBindings: ICheckMarkerFunctionBindings): Promise<void> {
+
+    const validatedCheck = this.findValidatedCheck(functionBindings.receivedCheckTable)
+    const markingData = await this.validateData(functionBindings, validatedCheck)
+    if (markingData === undefined) {
+      functionBindings.checkNotificationQueue.push({
+        checkCode: 'checkCode',
+        type: 'unmarkable'
+      })
+      return
+    }
+    const results = this.markCheck(markingData)
+    await this.persistMark(results, validatedCheck)
+    functionBindings.checkNotificationQueue.push({
+      checkCode: validatedCheck.RowKey,
+      type: 'marked'
+    })
+  }
+
+  private async validateData (functionBindings: ICheckMarkerFunctionBindings, validatedCheck: ValidatedCheck): Promise<MarkingData | void> {
+    if (RA.isEmptyString(validatedCheck.answers)) {
+      await this.updateReceivedCheckWithMarkingError(validatedCheck, 'answers property not populated')
+      return
+    }
+
+    let parsedAnswersJson: any
+    try {
+      parsedAnswersJson = JSON.parse(validatedCheck.answers)
+    } catch (error) {
+      await this.updateReceivedCheckWithMarkingError(validatedCheck, 'answers data is not valid JSON')
+      return
+    }
+
+    if (!RA.isArray(parsedAnswersJson)) {
+      await this.updateReceivedCheckWithMarkingError(validatedCheck, 'answers data is not an array')
+      return
+    }
+
+    const checkCode = validatedCheck.RowKey
+    const rawCheckForm = await this._sqlService.getCheckFormDataByCheckCode(checkCode)
```

Comment:
what if RowKey is not present in the validatedCheck?
shouldn't we prevent this transaction to happen in that case, call `updateReceivedCheckWithMarkingError` and return?

Reply:
thats the whole point of the check validator - the marker only receives it once its been validated and doesn't repeat all the work that the validator has done.